### PR TITLE
Add diff() and round() methods for Index and a test for each

### LINKED ISSUE #19708: ENH: Index should have diff() method

### PR DESCRIPTION
**Pull request by GrammatikakisDimitris**
- [x] closes #19708
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -6756,6 +6756,41 @@ def infer_objects(self, copy: bool = True) -> Index:
             result._references.add_index_reference(result)
         return result
 
+    def diff(self, periods: int = 1):
+        """
+        Computes the difference between consecutive values in the Index object.
+        If periods is greater than 1, computes the difference between values that
+        are `periods` number of positions apart.
+
+        Parameters:
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6765]: Looks the formatting here is off

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -1270,6 +1270,24 @@ def test_sortlevel_na_position(self):
         expected = Index([np.nan, 1])
         tm.assert_index_equal(result, expected)
 
+    def test_index_diff(self):
+        # GH#19708
+        idx = Index([10, 20, 30, 40, 50])
+        diffs = idx.diff()
+        diffs_period2 = idx.diff(periods=2)
```
Comment (mroeschke) [inline pandas/tests/indexes/test_base.py:1277]: Could you use `pytest.mark.parameterize` to test different periods?

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -1270,6 +1270,24 @@ def test_sortlevel_na_position(self):
...
+
+        assert diffs.equals(Index([np.nan, 10, 10, 10, 10]))
```
Comment (mroeschke) [inline pandas/tests/indexes/test_base.py:1279]: Could you make an `expected = ` variable and then call `tm.assert_index_equal(result, expected)`?

File: pandas/tests/indexes/test_base.py (hunk 2 above)
Comment (GrammatikakisDimitris) [reply pandas/tests/indexes/test_base.py:1277]: Would you like pytest.mark.parametrize for the round_test as well?

Comment (mroeschke) [reply pandas/tests/indexes/test_base.py:1277]: Yes please

File: pandas/tests/indexes/test_base.py [hunk 4]
```diff
@@ -1270,6 +1270,36 @@ def test_sortlevel_na_position(self):
         expected = Index([np.nan, 1])
         tm.assert_index_equal(result, expected)
 
+    @pytest.mark.parametrize(
+        "periods, expected",
+        [
+            (1, Index([np.nan, 10, 10, 10, 10])),
```
Comment (mroeschke) [inline pandas/tests/indexes/test_base.py:1276]: Could you move the `Index` call to the body of the test (same as below)

**Comment by GrammatikakisDimitris** [comment]
Forgot to pass parameters periods and decimals in the call of the functions, i will fix this.

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -6756,6 +6756,45 @@ def infer_objects(self, copy: bool = True) -> Index:
...
+        ----------
+        periods : int, optional
+            The number of positions between the current and previous
+            value to compute the difference with. Default is 1.
+
+        Returns
+        -------
+        Index
+            A new Index object with the computed differences.
+
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6776]: Could you add an `Examples` section. Same as below

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -6756,6 +6756,80 @@ def infer_objects(self, copy: bool = True) -> Index:
...
+        Example
+        -------
+        >>> import pandas as pd
+        >>> idx = pd.Index([10, 20, 30, 40, 50])
+        >>> diff_idx = idx.diff()
+        >>> print(diff_idx)
+        Index([nan, 10.0, 10.0, 10.0, 10.0], dtype='float64')
+
+        This example creates a new `Index` object `idx` with values
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6785]: The rest of this explanation is not needed

File: pandas/core/indexes/base.py [hunk 7]
```diff
@@ -6756,6 +6756,80 @@ def infer_objects(self, copy: bool = True) -> Index:
...
+            Number of decimal places to round to. If decimals is negative,
+            it specifies the number of positions to the left of the decimal point.
+
+        Returns
+        -------
+        Index
+            A new Index with the rounded values.
+
+        Example
+        -------
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6815]:
```suggestion
        Examples
        --------
```

File: pandas/core/indexes/base.py [hunk 8]
```diff
@@ -6756,6 +6756,80 @@ def infer_objects(self, copy: bool = True) -> Index:
...
+        >>> import pandas as pd
+        >>> idx = pd.Index([10.1234, 20.5678, 30.9123, 40.4567, 50.7890])
+        >>> rounded_idx = idx.round(decimals=2)
+        >>> print(rounded_idx)
+        Index([10.12, 20.57, 30.91, 40.46, 50.79], dtype='float64')
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6820]: This explanation is not needed

File: pandas/core/indexes/base.py [hunk 9]
```diff
@@ -6756,6 +6756,80 @@ def infer_objects(self, copy: bool = True) -> Index:
...
+            The number of positions between the current and previous
+            value to compute the difference with. Default is 1.
+
+        Returns
+        -------
+        Index
+            A new Index object with the computed differences.
+
+        Example
+        -------
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6778]:
```suggestion
        Examples
        --------
```

**Review by mroeschke — COMMENTED** [review]
Looks good. Could you add a whatsnew entry in `v2.1.0.rst`?

File: doc/source/whatsnew/v2.1.0.rst [hunk 10]
```diff
@@ -67,6 +67,40 @@ Also, note that :meth:`Categorical.map` implicitly has had its ``na_action`` set
 This has been deprecated and will :meth:`Categorical.map` in the future change the default
 to ``na_action=None``, like for all the other array types.
 
+.. _whatsnew_210.enhancements.enhancement3:
+
+Index now has diff() and round() methods
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:72]: A one line mention under the `Other Enhancements` section is sufficient

Comment (GrammatikakisDimitris) [reply doc/source/whatsnew/v2.1.0.rst]: My mistake, this one should be fine

**Comment by GrammatikakisDimitris** [comment]
What should i do about the conflict?

**Comment by mroeschke** [comment]
The conflict should be resolved without removing someone else's contribution

**Comment by GrammatikakisDimitris** [comment]
Sorry for the mess, i pulled all the changes to fix the conflict. If you like i can open a new PR.

**Comment by mroeschke** [comment]
I think all you need to do is

```
git fetch upstream
git merge upstream/main
```

into this branch

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @GrammatikakisDimitris

**Pull request merged**